lib/web/conn_upgrade.go

```diff
@@ -236,35 +237,46 @@ func (conn *waitConn) Close() error {
 }
 
 type websocketALPNServerConn struct {
-	*websocket.Conn
+	net.Conn
 	readBuffer []byte
 	readError  error
 	readMutex  sync.Mutex
 	writeMutex sync.Mutex
-}
 
-func newWebSocketALPNServerConn(wsConn *websocket.Conn) *websocketALPNServerConn {
-	return &websocketALPNServerConn{
-		Conn: wsConn,
-	}
+	logContext context.Context
+	logger     *slog.Logger
 }
 
-func (c *websocketALPNServerConn) convertError(err error) error {
-	if websocket.IsCloseError(err,
-		websocket.CloseAbnormalClosure,
-		websocket.CloseGoingAway,
-		websocket.CloseNormalClosure,
-	) {
-		return io.EOF
+func newWebSocketALPNServerConn(ctx context.Context, wsConn *websocket.Conn, logger *slog.Logger) *websocketALPNServerConn {
+	return &websocketALPNServerConn{
+		// Here "github.com/gobwas/ws" is used on the raw net.Conn instead of
+		// gorilla's websocket.Conn to workaround an issue that websocket.Conn
+		// caches read error when websocketALPNServerConn is passed to a HTTP
+		// server and get hijacked for another upgrade. Note that client side's
+		// (api/client) websocket ALPN connection wrapper also uses
+		// "github.com/gobwas/ws".
+		Conn:       wsConn.NetConn(),
+		logContext: ctx,
+		logger:     logger.With(teleport.ComponentKey, teleport.Component(teleport.ComponentWeb, "alpnws")),
 	}
-	return err
 }
 
 func (c *websocketALPNServerConn) Read(b []byte) (int, error) {
 	c.readMutex.Lock()
 	defer c.readMutex.Unlock()
 
 	n, err := c.readLocked(b)
+
+	// Timeout errors can be temporary. For example, when this connection is
+	// passed to the kube TLS server, it may get "hijacked" again. During the
+	// hijack, the SetReadDeadline is called with a past timepoint to fail this
+	// Read so that the HTTP server's background read can be stopped. In such
+	// cases, return the original net.Error and clear the cached read error.
+	if netError, ok := trace.Unwrap(err).(net.Error); ok && netError.Timeout() {
```

Comment:
nit: `errors.As()`? This works only for one level of nested errors AFAIK.

Reply:
`trace.Unwrap` calls `OrigError` which should unwrap for trace errors at least. `net.Error` is an interface and i don't have a concrete struct type. I will leave this as it is.

Reply:
the lint doesn't like the current way. converted to `errors.As` 👍